source/mpi_insim_supg.cpp

```diff
@@ -14,6 +14,362 @@ namespace Fluid
     template <int dim>
     void SUPGInsIM<dim>::assemble(const bool use_nonzero_constraints)
     {
+        TimerOutput::Scope timer_section(timer, "Assemble system");
+
+      Tensor<1, dim> gravity;
+      for (unsigned int i = 0; i < dim; ++i)
+        gravity[i] = parameters.gravity[i];
+
+      system_matrix = 0;
+      Abs_A_matrix = 0;
+      schur_matrix = 0;
+      B2pp_matrix = 0;
+
+      system_rhs = 0;
+
+      FEValues<dim> fe_values(fe,
+                              volume_quad_formula,
+                              update_values | update_quadrature_points |
+                                update_JxW_values | update_gradients);
+      FEFaceValues<dim> fe_face_values(fe,
+                                       face_quad_formula,
+                                       update_values | update_normal_vectors |
+                                         update_quadrature_points |
+                                         update_JxW_values);
+
+      const unsigned int dofs_per_cell = fe.dofs_per_cell;
+      const unsigned int u_dofs = fe.base_element(0).dofs_per_cell;
+      const unsigned int p_dofs = fe.base_element(1).dofs_per_cell;
+      const unsigned int n_q_points = volume_quad_formula.size();
+      const unsigned int n_face_q_points = face_quad_formula.size();
+
+      AssertThrow(u_dofs * dim + p_dofs == dofs_per_cell,
+                  ExcMessage("Wrong partitioning of dofs!"));
+
+      const FEValuesExtractors::Vector velocities(0);
+      const FEValuesExtractors::Scalar pressure(dim);
+
+      FullMatrix<double> local_matrix(dofs_per_cell, dofs_per_cell);
+      Vector<double> local_rhs(dofs_per_cell);
+
+      std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+      // For the linearized system, we create temporary storage for current
+      // velocity
+      // and gradient, current pressure, and present velocity. In practice, they
+      // are
+      // all obtained through their shape functions at quadrature points.
+
+      std::vector<Tensor<1, dim>> current_velocity_values(n_q_points);
+      std::vector<Tensor<2, dim>> current_velocity_gradients(n_q_points);
+      std::vector<double> current_pressure_values(n_q_points);
+      std::vector<Tensor<1, dim>> current_pressure_gradients(n_q_points);
+      std::vector<Tensor<1, dim>> present_velocity_values(n_q_points);
+      std::vector<double> present_pressure_values(n_q_points);
+      std::vector<double> sigma_pml(n_q_points);
+      std::vector<Tensor<1, dim>> artificial_bf(n_q_points);
+      std::vector<Tensor<1, dim>> fsi_acc_values(n_q_points);
```

Comment:
For 63-71: I'm certain some of the variables you declare here are not used. Check and remove them.